Add module for transforming samples from one format to another

fstanis opened the pull request:
This adds a `formatTransform` util that allows transforming from AMP for web to another format, specifically email.

The idea is to transform existing samples, e.g. [amp-form](https://amp.dev/documentation/examples/components/amp-form/?format=websites) "on the fly" to valid AMP for Email format. This is still work in progress, so the class is not currently used in the website.

sebastianbenz commented on platform/lib/format-transform/index.js:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  transform(input, target) {
+    if (!this.transforms[target]) {
```
why not use supportsFormat?

sebastianbenz commented on platform/lib/format-transform/transforms.js:
```diff
@@ -0,0 +1,66 @@
+  return url;
+}
+
+const amp4emailTransform = {
```
Please move this into it's own file.

sebastianbenz commented on platform/lib/format-transform/transforms.js:
```diff
@@ -0,0 +1,66 @@
+    el.removeAttr('target');
+  },
+  'amp-list': (el) => {
+    el.removeAttr('reset-on-refresh');
```
These transformations are dangerous as they break functionality. Instead, I'd throw an exception and mark the file as not transformable.

sebastianbenz commented on platform/lib/format-transform/transforms.js:
```diff
@@ -0,0 +1,66 @@
+  'amp-state[src]': null,
+  'amp-img[srcset]': null,
+  '[\\[src\\]]': null,
+  'input[type=password]': null,
```
same as above, this is functionality breaking.

sebastianbenz commented on platform/lib/utils/cheerio.js:
```diff
@@ -0,0 +1,41 @@
+/**
```
rename this file to cheerioHelper.js

sebastianbenz commented on platform/lib/format-transform/index.js:
```diff
@@ -0,0 +1,92 @@
+      transform(el);
+    } else if (typeof transform === 'string') {
+      el.replaceWith(transform);
+    }
```
add an else block throwing an exception

sebastianbenz commented on platform/lib/format-transform/index.js:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    if (typeof transform === 'function') {
+      transform(el);
```
nit:  early exit

sebastianbenz commented on platform/lib/format-transform/index.js:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  removeElement_(el) {
+    if (el.parent().is('div')) {
```
Why do we need to check for a parent div?

fstanis commented:
Addressed comments, most notably, removed the logic for element removal which was functionality breaking. We'll take a different approach here in another PR, to add conditional blocks to examples.

Also changed transformation format to include validatorRuntime that should be used to validate the code post-transform.

sebastianbenz commented on platform/lib/format-transform/index.js:
```diff
@@ -0,0 +1,71 @@
+
+  transform(input, target) {
+    if (!this.supportsFormat(target)) {
+      return input;
```
Throw an exception instead. Otherwise it's not possible for the caller to figure out if the transformation worked.

sebastianbenz commented on platform/lib/format-transform/index.js:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  transformElement_(el, transform) {
+    if (transform === null) {
```
add a comment explaining that this is the delete element case

sebastianbenz commented on platform/lib/format-transform/formats/amp4email.js:
```diff
@@ -0,0 +1,54 @@
+const host = config.getHost(config.hosts.platform);
+
+function absoluteUrl(url) {
+  if (!url.match(/^https?:\/\//)) {
```
Use `new URL(url, host).toString()` instead

fstanis replied:
Well TIL. I moved this method to config, let me know if that doesn't make sense.

sebastianbenz commented on platform/lib/format-transform/formats/amp4email.js:
```diff
@@ -0,0 +1,54 @@
+    'form, a': (el) => {
+      el.removeAttr('target');
+    },
+    'amp-list': (el) => {
```
this is also a functionality breaking transformation I would not support.

fstanis commented:
@sebastianbenz can we merge this?
